# In this coding file we are going to manipulate the data from the running shoes webpages of runrepeat and runningwarehouse to create:

# 1. The csv file from webpage1 (runrepeat) shoes that match in terms of similarity with the shoes from webpage2 (runningwarehouse).

# 2. The csv file from webpage2 that aren't duplicated. 

# 3. The csv file that match the similarity between the titles of these two webpages in terms of percentage.

In [1]:
# we are going to import the library of pandas and read the two webpages files
import pandas as pd 

In [2]:
webpage1 = pd.read_csv("runrepeat.csv")
webpage1 = webpage1.drop(columns = "Unnamed: 0")
webpage1

Title  price  price_discount  price_discount_%  \
0     361 Degrees Sensation 4   150.0            63.0              53.0   
1        361 Degrees Strata 3   150.0            27.0              83.0   
2        361 Degrees Meraki 2   130.0            19.0              85.0   
3     APL Streamline Lab test   180.0           120.0              60.0   
4           APL TechLoom Wave    60.0            92.0              62.0   
...                        ...    ...             ...               ...   
1082         Xero Shoes Zelen   185.0             0.0               0.0   
1083           Xero Shoes HFS   180.0             0.0               0.0   
1084          Xero Shoes Prio   150.0             0.0               0.0   
1085    Xero Shoes Mesa Trail   160.0             0.0               0.0   
1086   Xero Shoes Speed Force    65.0             0.0               0.0   

                brand                 model  
0                 361  Degrees Sensation 4   
1                 361     Degrees Strata 3   
2                 361     Degrees Meraki 2   
3     APL Performance  Streamline Lab test   
4     APL Performance        TechLoom Wave   
...               ...                   ...  
1082       Xero Shoes                Zelen   
1083       Xero Shoes                  HFS   
1084       Xero Shoes                 Prio   
1085       Xero Shoes           Mesa Trail   
1086       Xero Shoes          Speed Force   

[1087 rows x 6 columns]

In [3]:
webpage2 = pd.read_csv("running_warehouse.csv")
webpage2 = webpage2.drop(columns = "Unnamed: 0")
webpage2

title         brand  \
0             adidas adizero Prime X Strung        adidas   
1                adidas adizero Adios Pro 3        adidas   
2                adidas adizero Adios Pro 3        adidas   
3                adidas adizero Adios Pro 3        adidas   
4               adidas adizero Takumi Sen 9        adidas   
...                                     ...           ...   
996   Under Armour Centric Grip Throw Shoes  Under Armour   
997     Under Armour Shakedown Elite Spikes  Under Armour   
998      Under Armour HOVR Shakedown Spikes  Under Armour   
999     Under Armour HOVR Skyline LJ Spikes  Under Armour   
1000    Under Armour Kick Distance 4 Spikes  Under Armour   

                          model   price  discount  discount_%     sex  \
0        adizero Prime X Strung  299.95      0.00        0.00  unisex   
1           adizero Adios Pro 3  250.00    175.00       70.00  unisex   
2           adizero Adios Pro 3  250.00    175.00       70.00  unisex   
3           adizero Adios Pro 3  250.00    175.00       70.00  unisex   
4          adizero Takumi Sen 9  179.95      0.00        0.00  unisex   
...                         ...     ...       ...         ...     ...   
996    Centric Grip Throw Shoes   89.95      0.00        0.00  unisex   
997      Shakedown Elite Spikes  149.95      0.00        0.00  unisex   
998       HOVR Shakedown Spikes  119.95      0.00        0.00  unisex   
999      HOVR Skyline LJ Spikes   99.95      0.00        0.00  unisex   
1000     Kick Distance 4 Spikes   70.00     39.88       56.97  unisex   

                colour cushioning stability  \
0           Pulse Mint    Maximal   Neutral   
1       White Tint/Blk       High   Neutral   
2       Pulse Mint/Blk       High   Neutral   
3      Solar Red/Metal       High   Neutral   
4           Pulse Mint     Medium   Neutral   
...                ...        ...       ...   
996              Black        NaN       NaN   
997              White        NaN       NaN   
998          Halo Gray        NaN       NaN   
999               Gray        NaN       NaN   
1000       White/Bk/Mt        NaN       NaN   

                                       best_use_surface  \
0     Fast Runs,Cushioned Comfort,Racing,Roads and P...   
1                             Racing,Roads and Pavement   
2                             Racing,Roads and Pavement   
3                             Racing,Roads and Pavement   
4                             Racing,Roads and Pavement   
...                                                 ...   
996                      Shot Put/Discus/Hammer,Asphalt   
997                  800-1500m,Distance (3k-10k),Rubber   
998     800-1500m,Distance (3k-10k),Steeplechase,Rubber   
999                                                 NaN   
1000                           Distance (3k-10k),Rubber   

                                                midsole  \
0     Lightstrike Profoam compound offers high rebou...   
1     LightstrikePROis a lightweight midsole compoun...   
2     LightstrikePROis a lightweight midsole compoun...   
3     LightstrikePROis a lightweight midsole compoun...   
4     Lightstrike Prois a lightweight midsole compou...   
...                                                 ...   
996   HOVR Cushioning System includes a soft UA HOVR...   
997   Pelletized Pebax Foamin the midsole creates a ...   
998   HOVR Cushioning System includes a soft UA HOVR...   
999   HOVR Cushioning System includes a soft UA HOVR...   
1000  Micro-G Cushioning creates a balance of cushio...   

                                                outsole  \
0     Continental Rubber offers flexible durability ...   
1     Rubberin the forefootoffers flexible durabilit...   
2     Rubberin the forefootoffers flexible durabilit...   
3     Rubberin the forefootoffers flexible durabilit...   
4     Continental Rubberin the forefootoffers maximu...   
...                                                 ...   
996   Durable Rubber provides tr

# Identifying repeated values (entities) on these datasets, we need them unique on the webpage2 to compare the shoes with the weboage1 prices on SQL.

In [4]:
# with this method we can identify the duplicated values
webpage2["title"].duplicated()

0       False
1       False
2        True
3        True
4       False
        ...  
996     False
997     False
998     False
999     False
1000    False
Name: title, Length: 1001, dtype: bool

**With the sum() method we can count the True values only, because False values are 0 and True 1.**

In [5]:
# we do have more than a half of duplicated shoes here on running warehouse
webpage2["title"].duplicated().sum()

660

On the webpage2 of running warehouse many of the pair of shoes are duplicated in their title but there are some differences regarding their features of these duplicated shoes, sometimes it is because of sex, sometimes because of the colour or the appearance of discount etc. However, we are going to create two datasets with unique titles on webpage2 **(runninwarehouse)** to compare the prices of them with the most similar or equal shoes of webpage1 **(runrepeat)** with sql. 

**NOTE: the differences in price are very similar and mostly the same on this webpage2 with those duplicated titles (enitites) but despite it is not exactcly the same pair of shoes we are going to count them as it were because we want to focus on prices for the sake of this project. Later I will clean the data of running warehouse and update my statistical analysis and insights because we can't work and analyse the data with duplicated values**

In [6]:
# as you can see, regarding complete and exactly equal rows we have 33 duplicated observations
# I need to clean this data set of the running warehouse later.
webpage2.duplicated().sum()

33

In [7]:
# we don't have duplicates here on this webpage of runrepeat
webpage1["Title"].duplicated().sum()

0

In [8]:
# these are the duplicated rows in every feature but i am only displaying the title here
webpage2.loc[webpage2.duplicated(), ["title"]]

title
129                       ASICS Novablast 3
134                     ASICS Gel Nimbus 25
157                    ASICS Gel Cumulus 25
178                     ASICS Gel Kayano 29
205                        ASICS SUPERBLAST
234                      ASICS Noosa Tri 14
255                         Brooks Ghost 15
273                   Brooks Hyperion Tempo
285                     Brooks Hyperion Max
292                Brooks Adrenaline GTS 22
300                      Brooks Glycerin 20
320                      Brooks Cascadia 16
372                          HOKA Clifton 9
383                            HOKA Bondi 8
405                             HOKA Mach 5
419                            HOKA Arahi 6
428                           HOKA Rincon 3
455                       HOKA Challenger 7
532  New Balance FuelCell SuperComp Trainer
542           New Balance FuelCell Rebel v3
551        New Balance Fresh Foam X More v4
556       New Balance Fresh Foam X 1080 v12
580        New Balance Fresh Foam X 860 v13
591        New Balance Fresh Foam X 880 v13
782               Salomon S-Lab Pulsar 2 SG
783                  Salomon S-Lab Pulsar 2
784                    Salomon S-Lab Pulsar
826               Saucony Endorphin Speed 3
856                      Saucony Kinvara 14
872                         Saucony Ride 16
886                          Saucony Tempus
903                    Saucony Peregrine 13
926                        Saucony Guide 16

In [9]:
webpage2.shape

(1001, 15)

In [10]:
webpage2["title"].drop_duplicates().shape

(341,)

In [11]:
webpage1["Title"].shape

(1087,)

In [12]:
webpage2.drop_duplicates().shape

(968, 15)

In [13]:
# update the webpage2 to non duplicated values
webpage2 = webpage2.drop_duplicates(subset = ["title"])

# Webpage2 has 342 unique rows (shoes) and webpage1 1087, we want to compare how many pair of shoes from webpage1 are equal or similar to webpage2 regarding their titles. Remember the fuzz library for later.

In [14]:
# With this library we can use a similarity measure between two strings from different variables
from thefuzz import fuzz, process

In [15]:
# here is an example
s1 = "just a text"
s2 = "just another text"

fuzz.ratio(s1, s2)

79

Remember the shapes of these two current datasets we are working with 

In [16]:
webpage1.shape

(1087, 6)

In [17]:
webpage2.shape

(341, 15)

In [18]:
# as we selected non duplicated values from a dataframe with a thousand collumns
# we need to reset the index.
webpage2.reset_index(drop = True)

title         brand  \
0            adidas adizero Prime X Strung        adidas   
1               adidas adizero Adios Pro 3        adidas   
2              adidas adizero Takumi Sen 9        adidas   
3                 adidas adizero Boston 11        adidas   
4                   adidas adizero Adios 7        adidas   
..                                     ...           ...   
336  Under Armour Centric Grip Throw Shoes  Under Armour   
337    Under Armour Shakedown Elite Spikes  Under Armour   
338     Under Armour HOVR Shakedown Spikes  Under Armour   
339    Under Armour HOVR Skyline LJ Spikes  Under Armour   
340    Under Armour Kick Distance 4 Spikes  Under Armour   

                         model   price  discount  discount_%     sex  \
0       adizero Prime X Strung  299.95      0.00        0.00  unisex   
1          adizero Adios Pro 3  250.00    175.00       70.00  unisex   
2         adizero Takumi Sen 9  179.95      0.00        0.00  unisex   
3            adizero Boston 11  160.00    127.95       79.97   women   
4              adizero Adios 7  130.00    104.00       80.00   women   
..                         ...     ...       ...         ...     ...   
336   Centric Grip Throw Shoes   89.95      0.00        0.00  unisex   
337     Shakedown Elite Spikes  149.95      0.00        0.00  unisex   
338      HOVR Shakedown Spikes  119.95      0.00        0.00  unisex   
339     HOVR Skyline LJ Spikes   99.95      0.00        0.00  unisex   
340     Kick Distance 4 Spikes   70.00     39.88       56.97  unisex   

                 colour cushioning stability  \
0            Pulse Mint    Maximal   Neutral   
1        White Tint/Blk       High   Neutral   
2            Pulse Mint     Medium   Neutral   
3      Pulse Mint/White       High   Neutral   
4     Pulse Mint/Silver     Medium   Neutral   
..                  ...        ...       ...   
336               Black        NaN       NaN   
337               White        NaN       NaN   
338           Halo Gray        NaN       NaN   
339                Gray        NaN       NaN   
340         White/Bk/Mt        NaN       NaN   

                                      best_use_surface  \
0    Fast Runs,Cushioned Comfort,Racing,Roads and P...   
1                            Racing,Roads and Pavement   
2                            Racing,Roads and Pavement   
3              All Around,Fast Runs,Roads and Pavement   
4                         Fast Runs,Roads and Pavement   
..                                                 ...   
336                     Shot Put/Discus/Hammer,Asphalt   
337                 800-1500m,Distance (3k-10k),Rubber   
338    800-1500m,Distance (3k-10k),Steeplechase,Rubber   
339                                                NaN   
340                           Distance (3k-10k),Rubber   

                                               midsole  \
0    Lightstrike Profoam compound offers high rebou...   
1    LightstrikePROis a lightweight midsole compoun...   
2    Lightstrike Prois a lightweight midsole compou...   
3    LightStrike Prois a lightweight midsole compou...   
4    Lightstrike Prois a lightweight midsole compou...   
..                                                 ...   
336  HOVR Cushioning System includes a soft UA HOVR...   
337  Pelletized Pebax Foamin the midsole creates a ...   
338  HOVR Cushioning System includes a soft UA HOVR...   
339  HOVR Cushioning System includes a soft UA HOVR...   
340  Micro-G Cushioning creates a balance of cushio...   

                                               outsole  \
0    Continental Rubber offers flexible durability ...   
1    Rubberin the forefootoffers flexible durabilit...   
2    Continental Rubberin the forefootoffers maximu...   
3    Continental Rubberoffers flexible durability w...   
4    Continental Rubber Outsolein the forefoot offe...   
..                                                 ...   
336  Durable Rubber provides traction and a smooth ...   
337  Full-length

In [19]:
# update the webpage2 index
webpage2 = webpage2.reset_index(drop = True)

# The index will be the id on SQL the unique values for the entity relationship diagram

In [20]:
len(webpage1["Title"])

1087

In [21]:
len(webpage2["title"])

341

In [22]:
# we have a nan value on webpage2 file, it works as a float and we can't use the fuzz ratio
# if we have nan values, in this case we saw the type of each value in the title.
# but we could use isnull() too.
for i in webpage2["title"]:
    print(type(i))

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class

In [23]:
# the third row had a nan value
webpage2.iloc[2] 

title                                     adidas adizero Takumi Sen 9
brand                                                          adidas
model                                            adizero Takumi Sen 9
price                                                          179.95
discount                                                          0.0
discount_%                                                        0.0
sex                                                            unisex
colour                                                     Pulse Mint
cushioning                                                     Medium
stability                                                     Neutral
best_use_surface                            Racing,Roads and Pavement
midsole             Lightstrike Prois a lightweight midsole compou...
outsole             Continental Rubberin the forefootoffers maximu...
upper               Primegreenmesh is lightweight, breathable, and...
link_shoes          

In [24]:
# updating the title that was a nan value
webpage2.loc[2, "title"] = "adidas" 

In [25]:
webpage2.iloc[2] 

title                                                          adidas
brand                                                          adidas
model                                            adizero Takumi Sen 9
price                                                          179.95
discount                                                          0.0
discount_%                                                        0.0
sex                                                            unisex
colour                                                     Pulse Mint
cushioning                                                     Medium
stability                                                     Neutral
best_use_surface                            Racing,Roads and Pavement
midsole             Lightstrike Prois a lightweight midsole compou...
outsole             Continental Rubberin the forefootoffers maximu...
upper               Primegreenmesh is lightweight, breathable, and...
link_shoes          

# Identifying the highest similarity between the titles of these two webpages.

In [26]:
# prepare the final list to build the dataframe
list_dictionaries = []

# iterate through the webpage2 with 341 pair of shoes
for i in webpage2["title"]:
    
    #select the highest ratio
    highest_ratio = 0
    
    # prepare a dictionary to append into the list_dictionaries
    dictionary = {}
    
    # iterate through the webpage1 to compare each similarity ratio
    # with the current observation of webpage2
    for j in webpage1["Title"]:
        
        # get the current ratio, this is where we use the fuzz library
        current_ratio = fuzz.ratio(i, j)
        
        # update to the highest ratio along with the title of webpage1, select the highest
        if current_ratio > highest_ratio:
            
            highest_ratio = current_ratio
            
            # append the current title into the dictionary after setting the highest ratio
            dictionary["webpage1"] = j
    
    # set the current title for webpage2
    dictionary["webpage2"] = i
    
    # set the highest similarity ratio for both titles of the webpages 1-2 
    dictionary["highest_similarity"] = highest_ratio
    
    # append the dictionary corresponding to the webpage2 iteration
    list_dictionaries.append(dictionary)

In [27]:
equal_titles = pd.DataFrame(list_dictionaries)

In [28]:
equal_titles

webpage1                               webpage2  \
0    Adidas Adizero Prime X Strung           adidas adizero Prime X Strung   
1       Adidas Adizero Adios Pro 3              adidas adizero Adios Pro 3   
2                        Adidas 4D                                  adidas   
3         Adidas Adizero Boston 11                adidas adizero Boston 11   
4           Adidas Adizero Adios 5                  adidas adizero Adios 7   
..                              ...                                    ...   
336          Under Armour Fat Tire   Under Armour Centric Grip Throw Shoes   
337    Under Armour Charged Bandit     Under Armour Shakedown Elite Spikes   
338      Under Armour HOVR Sonic 4      Under Armour HOVR Shakedown Spikes   
339      Under Armour HOVR Sonic 4     Under Armour HOVR Skyline LJ Spikes   
340      Under Armour HOVR Sonic 4     Under Armour Kick Distance 4 Spikes   

     highest_similarity  
0                    92  
1                    91  
2                    62  
3                    90  
4                    84  
..                  ...  
336                  61  
337                  67  
338                  73  
339                  72  
340                  62  

[341 rows x 3 columns]

**As you can see we have some repeated values on webpage1 because we want to compare each pair of shoes from the webpage2 with the most similar pair of shoes of the webpage1 even if these repeat.**

In [29]:
equal_titles["webpage1"].duplicated().sum()

102

In [30]:
equal_titles["webpage2"].duplicated().sum()

0

**Now it is time to select all of those 341 values from the webpage1 runrepeat**

In [31]:
# if we are going to merge, we need to set the same name for the two columns
# we are going to compare with each other.
equal_titles = equal_titles.rename(columns = {"webpage1": "Title"})

In [32]:
# I applied the merge() function to match the titles of the shoes
# from the webpage1 along with the features of this webpage1 like 
# price, discount etc.
equal_titles.merge(webpage1, how = "inner")

Title                               webpage2  \
0    Adidas Adizero Prime X Strung           adidas adizero Prime X Strung   
1       Adidas Adizero Adios Pro 3              adidas adizero Adios Pro 3   
2                        Adidas 4D                                  adidas   
3         Adidas Adizero Boston 11                adidas adizero Boston 11   
4           Adidas Adizero Adios 5                  adidas adizero Adios 7   
..                              ...                                    ...   
336          Under Armour Fat Tire   Under Armour Centric Grip Throw Shoes   
337    Under Armour Charged Bandit     Under Armour Shakedown Elite Spikes   
338      Under Armour HOVR Sonic 4      Under Armour HOVR Shakedown Spikes   
339      Under Armour HOVR Sonic 4     Under Armour HOVR Skyline LJ Spikes   
340      Under Armour HOVR Sonic 4     Under Armour Kick Distance 4 Spikes   

     highest_similarity  price  price_discount  price_discount_%  \
0                    92  100.0           230.0              23.0   
1                    91   80.0           124.0              50.0   
2                    62  170.0             0.0               0.0   
3                    90  180.0            79.0              51.0   
4                    84  150.0            52.0              63.0   
..                  ...    ...             ...               ...   
336                  61  150.0             0.0               0.0   
337                  67   80.0            50.0              50.0   
338                  73  150.0            48.0              56.0   
339                  72  150.0            48.0              56.0   
340                  62  150.0            48.0              56.0   

            brand                    model  
0          Adidas  Adizero Prime X Strung   
1          Adidas     Adizero Adios Pro 3   
2          Adidas                      4D   
3          Adidas       Adizero Boston 11   
4          Adidas         Adizero Adios 5   
..            ...                      ...  
336  Under Armour                Fat Tire   
337  Under Armour          Charged Bandit   
338  Under Armour            HOVR Sonic 4   
339  Under Armour            HOVR Sonic 4   
340  Under Armour            HOVR Sonic 4   

[341 rows x 8 columns]

**We are going to save the three dataframes we wanted to create**

In [33]:
webpage2.to_csv("webpage2_warehouse.csv")

In [34]:
# we are going to select only the relevant columns for the webpage1, runrepeat
webpage1 = equal_titles.merge(webpage1, how = "inner")[["Title", "brand", "model","price", 
                                                        "price_discount", "price_discount_%"]]

In [35]:
webpage1.to_csv("webpage1_runrepeat.csv")

In [36]:
equal_titles.to_csv("equal_titles.csv")